ignite/metrics/precision.py

```diff
@@ -15,28 +15,47 @@ class _BasePrecisionRecall(_BaseClassification):
     def __init__(
         self,
         output_transform: Callable = lambda x: x,
-        average: bool = False,
+        average: Optional[Union[bool, str]] = False,
         is_multilabel: bool = False,
         device: Union[str, torch.device] = torch.device("cpu"),
     ):
 
-        self._average = average
+        if not (average is None or isinstance(average, bool) or average in ["macro", "micro", "weighted", "samples"]):
+            raise ValueError(
+                "Argument average should be None or a boolean or one of values"
+                " 'macro', 'micro', 'weighted' and 'samples'."
+            )
+
+        if average is None:
+            self._average = False  # type: Union[bool, str]
+        elif average is True:
+            self._average = "macro"
+        else:
+            self._average = average
         self.eps = 1e-20
         self._updated = False
         super(_BasePrecisionRecall, self).__init__(
             output_transform=output_transform, is_multilabel=is_multilabel, device=device
         )
 
+    def _check_type(self, output: Sequence[torch.Tensor]) -> None:
+        super()._check_type(output)
+
+        if self._type in ["binary", "multiclass"] and self._average == "samples":
+            raise ValueError("Argument average='samples' is incompatible with binary and multiclass input data.")
+
     @reinit__is_reduced
     def reset(self) -> None:
-        self._true_positives = 0  # type: Union[int, torch.Tensor]
-        self._positives = 0  # type: Union[int, torch.Tensor]
-        self._updated = False
+        if self._average == "samples":
+            self._sum_samples_metric = 0  # type: Union[int, torch.Tensor]
```

Comment:
@sadra-barikbin you introduce here new accumulators: `_sum_samples_metric`, `_samples_cnt`, `_actual_positives` in addition to existing `_true_positives`, `_positives`. Can we just keep 3 instead, we can rename them to something abstract if names like `_true_positives` and `_positives` are inappropriate.

Reply:
Yes. For example `nominator`, `denominator` and `weight`. It hurts readability, though. Not only the type but also the semantic of `_sum_samples_metric` is different with  that of `_true_positives` or `_positives`. The former is a float value representing the summation of computed metric over samples and the latter is the number of samples.

Reply:
I added comment to `reset` and `compute` functions which are also common between precision and recall. Is it ok?